src/Raven.Server/Documents/Indexes/IndexStore.cs

```diff
@@ -594,136 +764,104 @@ public void Dispose()
                 exceptionAggregator.Execute(index.Dispose);
             });
 
+            _serverStore.Cluster.DatabaseChanged -= OnDatabaseChanged;
+
             exceptionAggregator.ThrowIfNeeded();
         }
 
-        private int ResetIndexInternal(Index index)
+        private async Task<long> ResetIndexInternal(Index index)
         {
-            lock (_indexAndTransformerLocker)
-            {
-                DeleteIndex(index.IndexId);
-                return CreateIndex(index.Definition);
-            }
+            await DeleteIndex(index.Etag);
+            return await CreateIndex(index.Definition);
         }
 
-        private void OpenIndexes()
+        private void OpenIndexes(DatabaseRecord record)
         {
             if (_documentDatabase.Configuration.Indexing.RunInMemory)
                 return;
 
-            lock (_indexAndTransformerLocker)
+            lock (_locker)
             {
-                OpenIndexesFromDirectory(_documentDatabase.Configuration.Indexing.StoragePath);
-
-                if (_documentDatabase.Configuration.Indexing.AdditionalStoragePaths != null)
-                {
-                    foreach (var path in _documentDatabase.Configuration.Indexing.AdditionalStoragePaths)
-                        OpenIndexesFromDirectory(path);
-                }
+                // apply renames
+                OpenIndexesFromRecord(_documentDatabase.Configuration.Indexing.StoragePath, record);
             }
         }
 
-        private void OpenIndexesFromDirectory(PathSetting path)
+        private void OpenIndexesFromRecord(PathSetting path, DatabaseRecord record)
         {
-            if (Directory.Exists(path.FullPath) == false)
-                return;
-
             if (_logger.IsInfoEnabled)
-                _logger.Info($"Starting to load indexes from {path}");
+                _logger.Info($"Starting to load indexes from record");
+
+            List<Exception> exceptions = null;
+            if (_documentDatabase.Configuration.Core.ThrowIfAnyIndexOrTransformerCouldNotBeOpened)
+                exceptions = new List<Exception>();
 
-            var indexes = new SortedList<int, Tuple<string, string>>();
-            foreach (var indexDirectory in new DirectoryInfo(path.FullPath).GetDirectories())
+            foreach (var kvp in record.Indexes)
```

Comment:
Another issue that we have to deal with, and why we probably have to retain our local storage option is the customization of index paths, either globally or for a particular node.

We need to support a more complex system, and in this case, just storing it all directly, instead of probing the file system will be better.